### ISSUE TITLE
0.227 build failure with gcc-11 snapshot (Fedora rawhide)

### ISSUE THREAD

**Issue by belegdol** [mamedev/mame#7616]
Hi,
mame-0.227 has failed to build on Fedora rawhide:
https://kojipkgs.fedoraproject.org//work/tasks/4782/58664782/build.log
```
Compiling src/emu/emumem_hedr0.cpp...
g++    -MMD -MP -MP -DPTR64=1 -DNDEBUG -DCRLF=2 -DLSB_FIRST -DXMD_H -DFLAC__NO_DLL -DNATIVE_DRC=drcbe_x64 -DLUA_COMPAT_ALL -DLUA_COMPAT_5_1 -DLUA_COMPAT_5_2 -I"../../../../../src/osd" -I"../../../../../src/emu" -I"../../../../../src/lib" -I"../../../../../src/lib/util" -I"../../../../../3rdparty" -I"../../../../../3rdparty/sol2" -I"../../../../generated/emu" -I"../../../../generated/emu/layout" -I"../../../../../3rdparty/asio/include" -I"../../../../../scripts/src"  -m64 -pipe -O2 -fno-strict-aliasing -O2 -fexceptions -g1 -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wno-unknown-pragmas -Wall -Wcast-align -Wformat-security -Wundef -Wwrite-strings -Wno-conversion -Wno-sign-compare -Wno-error=deprecated-declarations -Wno-unused-result -Wno-array-bounds -Wno-format-overflow -Wno-stringop-truncation -Wno-stringop-overflow -Wno-return-local-addr -m64 -std=c++17 -Woverloaded-virtual -Wimplicit-fallthrough -Wno-class-memaccess -Wsuggest-override -flifetime-dse=1 -include ../../../../linux_gcc/obj/x64/Release/emu.h -o "../../../../linux_gcc/obj/x64/Release/src/emu/emumem_hedr0.o" -c "../../../../../src/emu/emumem_hedr0.cpp"
In file included from ../../../../../src/emu/emumem_hedr0.cpp:5:
../../../../../src/emu/emumem_hedr.ipp: In instantiation of 'void handler_entry_read_dispatch<HighBits, Width, AddrShift, Endian>::populate_nomirror_subdispatch(offs_t, offs_t, offs_t, offs_t, offs_t, handler_entry_read<Width, AddrShift, Endian>*) [with int HighBits = 0; int Width = 0; int AddrShift = 1; endianness_t Endian = ENDIANNESS_LITTLE; offs_t = unsigned int]':
../../../../../src/emu/emumem_hedr0.cpp:8:16:   required from here
../../../../../src/emu/emumem_hedr.ipp:169:118: error: 'address_space* handler_entry::m_space' is protected within this context
  169 |                 auto subdispatch = new handler_entry_read_dispatch<LowBits, Width, AddrShift, Endian>(handler_entry::m_space, m_u_ranges[entry], cur);
      |                                                                                                                      ^~~~~~~
In file included from ../../../../../src/emu/emu.h:51:
../../../../../src/emu/emumem.h:550:24: note: declared protected here
  550 |         address_space *m_space;
      |                        ^~~~~~~
In file included from ../../../../../src/emu/emumem_hedr0.cpp:5:
../../../../../src/emu/emumem_hedr.ipp: In instantiation of 'void handler_entry_read_dispatch<HighBits, Width, AddrShift, Endian>::populate_mirror_subdispatch(offs_t, offs_t, offs_t, offs_t, offs_t, offs_t, handler_entry_read<Width, AddrShift, Endian>*) [with int HighBits = 0; int Width = 0; int AddrShift = 1; endianness_t Endian = ENDIANNESS_LITTLE; offs_t = unsigned int]':
../../../../../src/emu/emumem_hedr0.cpp:8:16:   required from here
../../../../../src/emu/emumem_hedr.ipp:235:118: error: 'address_space* handler_entry::m_space' is protected within this context
  235 |                 auto subdispatch = new handler_entry_read_dispatch<LowBits, Width, AddrShift, Endian>(handler_entry::m_space, m_u_ranges[entry], cur);
      |                                                                                                                      ^~~~~~~
In file included from ../../../../../src/emu/emu.h:51:
../../../../../src/emu/emumem.h:550:24: note: declared protected here
  550 |         address_space *m_space;
      |                        ^~~~~~~
In file included from ../../../../../src/emu/emumem_hedr0.cpp:5:
../../../../../src/emu/emumem_hedr.ipp: In instantiation of 'void handler_entry_read_dispatch<HighBits, Width, AddrShift, Endian>::populate_mismatched_nomirror_subdispatch(offs_t, offs_t, offs_t, offs_t, offs_t, const memory_units_descriptor<Width, AddrShift, Endian>&, u8, std::vector<typename handler_entry_read<Width, AddrShift, Endian>::mapping>&) [with int HighBits = 0; int Width = 0; int AddrShift = 1; endianness_t Endian = ENDIANNESS_LITTLE; offs_t = unsigned int; u8 = unsigned char; typename handler_entry_read<Width, AddrShift, Endian>::mapping = handler_entry_read<0, 1, ENDIANNESS_LITTLE>::mapping]':
../../../../../src/emu/emumem_hedr0.cpp:8:16:   required from here
../../../../../src/emu/emumem_hedr.ipp:303:118: error: 'address_space* handler_entry::m_space' is protected within this context
  303 |                 auto subdispatch = new handler_entry_read_dispatch<LowBits, Width, AddrShift, Endian>(handler_entry::m_space, m_u_ranges[entry], cur);
      |                                                                                                                      ^~~~~~~
In file included from ../../../../../src/emu/emu.h:51:
../../../../../src/emu/emumem.h:550:24: note: declared protected here
  550 |         address_space *m_space;
      |                        ^~~~~~~
In file included from ../../../../../src/emu/emumem_hedr0.cpp:5:
../../../../../src/emu/emumem_hedr.ipp: In instantiation of 'void handler_entry_read_dispatch<HighBits, Width, AddrShift, Endian>::populate_mismatched_mirror_subdispatch(offs_t, offs_t, offs_t, offs_t, offs_t, offs_t, const memory_units_descriptor<Width, AddrShift, Endian>&, std::vector<typename handler_entry_read<Width, AddrShift, Endian>::mapping>&) [with int HighBits = 0; int Width = 0; int AddrShift = 1; endianness_t Endian = ENDIANNESS_LITTLE; offs_t = unsigned int; typename handler_entry_read<Width, AddrShift, Endian>::mapping = handler_entry_read<0, 1, ENDIANNESS_LITTLE>::mapping]':
../../../../../src/emu/emumem_hedr0.cpp:8:16:   required from here
../../../../../src/emu/emumem_hedr.ipp:375:118: error: 'address_space* handler_entry::m_space' is protected within this context
```
This goes on for quite a while. I am not sure if it is a real problem or gcc issue. Thanks for checking.

**Comment by rb6502** [comment]
I would assume since this works on Clang and GCC 9 and 10 that it's probably a compiler bug, but I'll tag @galibert to make sure.

**Comment by belegdol** [comment]
Got it. I am happy to report it to gcc maintainers if it is a compiler bug, but they are probably going to ask for a reproducer snippet.

**Comment by kaincenteno** [comment]
Wasn't unsure if I should open a new issue but the same version is also failing to compile on Ubuntu 20.10 with gcc 10.2.0

```
Compiling src/devices/bus/nubus/nubus.cpp...
Compiling src/devices/bus/nubus/nubus_48gc.cpp...
Compiling src/devices/bus/nubus/nubus_asntmc3b.cpp...
Compiling src/devices/bus/nubus/nubus_cb264.cpp...
Compiling src/devices/bus/nubus/nubus_image.cpp...
Compiling src/devices/bus/nubus/nubus_m2hires.cpp...
Compiling src/devices/bus/nubus/nubus_m2video.cpp...
Compiling src/devices/bus/nubus/nubus_radiustpd.cpp...
In file included from /usr/include/string.h:519,
                 from /usr/include/c++/10/cstring:42,
                 from ../../../../../src/emu/emucore.h:20,
                 from ../../../../../src/emu/emu.h:30:
In function ‘char* strcat(char*, const char*)’,
    inlined from ‘void nubus_image_device::file_cmd_w(uint32_t)’ at ../../../../../src/devices/bus/nubus/nubus_image.cpp:261:10:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:129:33: error: ‘char* __builtin___strcat_chk(char*, const char*, long unsigned int)’ accessing 129 or more bytes at offsets 1012 and 884 may overlap 1 byte at offset 1012 [-Werror=restrict]
  129 |   return __builtin___strcat_chk (__dest, __src, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiling src/devices/bus/nubus/nubus_spec8.cpp...
Compiling src/devices/bus/nubus/nubus_specpdq.cpp...
cc1plus: all warnings being treated as errors
make[2]: *** [optional.make:18796: ../../../../linux_gcc/obj/x64/Release/src/devices/bus/nubus/nubus_image.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:100: optional] Error 2
make: *** [makefile:1379: linux_x64] Error 2
```

**Comment by kaincenteno** [comment]
ignore my last comment, followed what was on https://github.com/mamedev/mame/pull/7613 to fix it

**Comment by galibert** [comment]
Not being allowed to access a protected member from a derived class is a new one for me...

**Comment by galibert** [comment]
A shortened version that shows the problem is on
https://gcc.godbolt.org/z/hPbch4

You can see it works with every version but gcc trunk.

It's parts of the standard that are so kinky I'm not sure of anything
though.  Vas, do you have an idea?

  OG.

**Comment by cuavas** [comment]
If you change `A::var0++` to `this->var0++` it compiles correctly.  Why are you using `A::var0` syntax for member anyway?  using `this->var0` is enough to make it into an argument-dependent name and defer lookup.

**Comment by cuavas** [comment]
Here’s a minimal case that fails:
```
class A {
public:
    A() = default;

protected:
    int var0 = 0;
};

template<int a> class B : public A {
public:
    using A::A;
};

template<int a, int b> class C : public B<a> {
public:
    using B<a>::B;

    void g();
};

template<int a, int b> void C<a, b>::g()
{
    A::var0++;
}

template class C<0, 0>;
```

Interestingly, it compiles when you inline the body of `C::g` like this:
```
class A {
public:
    A() = default;

protected:
    int var0 = 0;
};

template<int a> class B : public A {
public:
    using A::A;
};

template<int a, int b> class C : public B<a> {
public:
    using B<a>::B;

    void g() { A::var0++; };
};

template class C<0, 0>;
```

Given that it’s behaving differently depending on whether or not the body of the function is inlined, I’d say this is almost definitely a compiler bug.

**Comment by galibert** [comment]
Why do I have to use anything in a clear case of inheritance anyway.  Also,
B<a>:: works too, but of course that's no the point.

  OG.

**Comment by cuavas** [comment]
Because the base is a template, and partial specialisation is a thing. A partial specialisation of the base could be declared later with different members. To prevent undesirable behaviour, the compiler doesn't search base templates for names that aren't argument-dependent. It has to know when it parses your template whether something is supposed to come from a templated base or from namespace scope.

TL;DR you have to because the alternative is worse.

**Comment by belegdol** [comment]
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98515

**Comment by belegdol** [comment]
@cuavas, with your changes mame compiles again. Thanks!
